api_log/models/api_log.py

```diff
@@ -0,0 +1,224 @@
+    _description = "Log for API"
+
+    # Request
+    request_url = fields.Char()
+    request_method = fields.Char()
+    request_headers = fields.Json()
+    request_body = fields.Binary(attachment=False)
+    request_date = fields.Datetime()
+    request_time = fields.Float()
+
+    # Response
+    response_status_code = fields.Integer()
+    response_headers = fields.Json()
+    response_body = fields.Binary(attachment=False)
+    response_date = fields.Datetime()
+    response_time = fields.Float()
+
+    stack_trace = fields.Text()
+
+    # Derived fields
+    name = fields.Char(compute="_compute_name", store=True)
+    time = fields.Float(compute="_compute_time", store=True)
+    request_preview = fields.Text(compute="_compute_request_preview")
+    response_preview = fields.Text(compute="_compute_response_preview")
+    request_b64 = fields.Binary(
+        string="Request Content", compute="_compute_request_b64"
+    )
+    response_b64 = fields.Binary(
+        string="Response Content", compute="_compute_response_b64"
+    )
+    request_headers_preview = fields.Text(compute="_compute_headers_preview")
+    response_headers_preview = fields.Text(compute="_compute_headers_preview")
+    request_content_type = fields.Char(
+        compute="_compute_request_headers_derived", store=True
+    )
+    request_content_length = fields.Integer(
+        compute="_compute_request_headers_derived", store=True
+    )
+    referrer = fields.Char(compute="_compute_request_headers_derived", store=True)
+    response_content_type = fields.Char(
+        compute="_compute_response_headers_derived", store=True
+    )
+    response_content_length = fields.Integer(
+        compute="_compute_response_headers_derived", store=True
+    )
+
+    @api.model
+    def _headers_hidden_keys(self):
+        """Header keys that should not be logged.
+
+        They might contains sensitive data.
+        """
+        return (
+            "Api-Key",
+            "Cookie",
+        )
+
+    @api.model
+    def _sanitize_headers_dict(self, headers_dict):
+        keys_to_hide = self._headers_hidden_keys()
+        for key in headers_dict:
+            if key in keys_to_hide:
+                headers_dict[key] = "<redacted>"
+        return headers_dict
+
+    @api.model
+    def _headers_to_dict(self, headers):
+        try:
+            headers_dict = {key: value for key, value in headers.items()}
+            return self._sanitize_headers_dict(headers_dict)
+        except AttributeError:
+            return {}
+
+    def _current_time(self):
+        return time.time_ns() / 1e9
+
+    @api.model
+    def _get_http_request(self, request):
+        return request.httprequest
+
+    @api.model
+    def _get_request_body(self, request):
+        """Take extra care with the request's body because it might get consumed."""
+        httprequest = self._get_http_request(request)
+        return httprequest.data
+
+    @api.model
+    def log_request(self, request):
```

Comment:
This should accept custom values. Eg:
```suggestion
    def log_request(self, request, **values):
        log_request_values = {...}
        log_request_values.update(values)
```
This way you don't have to override the api.log model every time you implement a glue module.
Plus, it should use `sudo` to create.

Reply:
> This way you don't have to override the api.log model every time you implement a glue module.

I prefer to override the method once than forcing the developer to remember to add the custom values each time this method is called; I've added them as suggested because they could be useful anyway.

>  Plus, it should use sudo to create.

I did let the caller choose if they wanted to use `sudo` or not, but now that you mention it I don't see why it should _not_ be used, so I added it here  as suggested and in other `write` calls.